weather_mv/setup.py

```diff
@@ -21,15 +32,98 @@
     "xarray",
     "google-cloud-bigquery",
     "pyparsing==2.4.2",  # Fix http2lib auth breakage
+    "cfgrib",
+    "netcdf4",
 ]
 
+
+# This class handles the pip install mechanism.
+class build(_build):  # pylint: disable=invalid-name
+    """A build command class that will be invoked during package install.
+    The package built using the current setup.py will be staged and later
+    installed in the worker using `pip install package'. This class will be
+    instantiated during install for this specific scenario and will trigger
+    running the custom commands specified.
+    """
+    sub_commands = _build.sub_commands + [('CustomCommands', None)]
+
+
+# Some custom command to run during setup. The command is not essential for this
+# workflow. It is used here as an example. Each command will spawn a child
+# process. Typically, these commands will include steps to install non-Python
+# packages. For instance, to install a C++-based library libjpeg62 the following
+# two commands will have to be added:
+#
+#     ['apt-get', 'update'],
+#     ['apt-get', '--assume-yes', 'install', 'libjpeg62'],
+#
+# First, note that there is no need to use the sudo command because the setup
+# script runs with appropriate access.
+# Second, if apt-get tool is used then the first command needs to be 'apt-get
+# update' so the tool refreshes itself and initializes links to download
+# repositories.  Without this initial step the other apt-get install commands
+# will fail with package not found errors. Note also --assume-yes option which
+# shortcuts the interactive confirmation.
+#
+# Note that in this example custom commands will run after installing required
+# packages. If you have a PyPI package that depends on one of the custom
+# commands, move installation of the dependent package to the list of custom
+# commands, e.g.:
+#
+#     ['pip', 'install', 'my_package'],
+#
+# TODO(BEAM-3237): Output from the custom commands are missing from the logs.
+# The output of custom commands (including failures) will be logged in the
+# worker-startup log.
+CUSTOM_COMMANDS = [
+    cmd.split() for cmd in [
+        'apt-get update',
+        'apt-get --assume-yes install libeccodes-dev'
+    ]
+]
+
+
+class CustomCommands(Command):
+    """A setuptools Command class able to run arbitrary commands."""
+
+    def initialize_options(self):
+        pass
+
+    def finalize_options(self):
+        pass
+
+    def RunCustomCommand(self, command_list):
+        print('Running command: %s' % command_list)
+        p = subprocess.Popen(
+            command_list,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT)
+        # Can use communicate(input='y\n'.encode()) if the command run requires
+        # some confirmation.
+        stdout_data, _ = p.communicate()
+        print('Command output: %s' % stdout_data)
```

Comment:
Should this use logging for consistency?

Reply:
I'm OK with using `print` statements in a custom command. It would be a bit tricky to setup the logger during the install process. Further, this code is copied from the Apache Beam docs. I'd rather keep it the same as much as possible for easier maintenance.